File: eight-queens.js
// EIGHT QUEENS MATRIX PROBLEM

/*
The eight queens puzzle is the problem of placing eight chess queens on an 8×8 chessboard so that no two queens threaten each other. Thus, a solution requires that no two queens share the same row, column, or diagonal.

Return an array/list with all the possible solutions (either in a 2D matrix each or other data structure with the positions on the board).

Be mindful not to exceed the maximum recursion depth in most JS/python envorinments. This is a long one, so don't worry if you need more than an hour.

Good luck!
*/

function validatePosition(arr, y) {
    // create slanted variables
    let rightUp = y;
    let leftUp = y;

    // iterate through the already placed queens array and check if the new pos is invalid
    for(let i = arr.length - 1; i >= 0; i--) {
        rightUp++;
        leftUp--;
        if(arr[i] === y || (rightUp < 8 && arr[i] === rightUp) || (leftUp >= 0 && arr[i] === leftUp)) {
            return false;
        }
    }
    // otherwise return true
    return true;
}

function eightQueens(positions = [], y = 0, results = [], limit = 1) {
    // return results when exceed the limit
    if(positions.length === 0 && y === limit) {
        return results;
    }

    // if the positions array is 8 long, add it to the results array
    if(positions.length === 8) {
        results.push([...positions]);

        // and continue search with new last position
        let newY = positions.pop() + 1;
        return eightQueens(positions, newY, results, limit);
    }

    // if the position is out of range roll back to last line
    if(y >= 8) {
        let newY = positions.pop() + 1;
        return eightQueens(positions, newY, results, limit);
    } else {
        // if the position in the line validates continue to next line, else try next position
        if(validatePosition(positions, y)) {
            positions.push(y);
            return eightQueens(positions, 0, results, limit);
        } else {
            return eightQueens(positions, y + 1, results, limit);
        }
    }
}

// result compiler function to circumvent exceeding max stack size issue
function compileResults() {
    let allConfigurations = [];

    for(let i = 0; i < 8; i++) {
        allConfigurations.push(...eightQueens([], i, [], i + 1));
    }

    return allConfigurations;
}

console.log(compileResults());

File: eight-queens.py
# EIGHT QUEENS MATRIX PROBLEM
def validate_position(positions, y):
    right = y
    left = y
    for pos in reversed(positions):
        right = right + 1
        left = left - 1
        if(pos == left or pos == y or pos == right):
            return False
    return True


def eight_queens(positions=[], y=0, results=[]):
    # final success base case
    if(len(positions) == 0 and y == 8):
        return results
    if(len(results) == 8):
        return results
    
    # new config found base case + continue search
    if(len(positions) == 8):
        results.append([pos for pos in positions])
        newY = positions.pop() + 1
        return eight_queens(positions, newY, results)

    # rollback base case
    if(y == 8):
        newY = positions.pop() + 1
        return eight_queens(positions, newY, results)
    else:
        if(validate_position(positions, y)):
            positions.append(y)
            # print(positions)
            return eight_queens(positions, 0, results)
        else:
            return eight_queens(positions, y + 1, results)


print(eight_queens())